### PR TITLE
NetReferences don't use GCHandles any more

### PR DESCRIPTION
**Pull request by devmil**
References are counted manually on .Net side
Behaviors now don't get an event on each NetReference create / delete but when an .Net object enters Native the first time and when it leaves the Native space

### CODE REVIEW

File: src/native/QmlNet/QmlNet/types/NetReference.cpp [hunk 1]
```diff
@@ -33,20 +26,17 @@ QSharedPointer<NetTypeInfo> NetReference::getTypeInfo()
 
 void NetReference::release()
```
Comment (pauldotknopf) [inline src/native/QmlNet/QmlNet/types/NetReference.cpp]: Do we still need this ```release``` method? Shouldn't we just be adding the defref stuff in the deconstructor?

File: src/net/Qml.Net/Internal/Types/NetReference.cs [hunk 2]
```diff
@@ -106,17 +107,84 @@ public static void ReleaseGCHandle(GCHandle handle, UInt64 objectId)
...
+                Counter = count;
+            }
+        }
+
+        private static Dictionary<UInt64, ObjectEntry> _ObjectIdObjectLookup = new Dictionary<ulong, ObjectEntry>();
+        private static object _LockObject = new object();
+
+        internal static bool TryGetObjectFor(UInt64 objectId, out object obj)
+        {
+            lock(_LockObject)
```
Comment (pauldotknopf) [inline src/net/Qml.Net/Internal/Types/NetReference.cs]: Do we need to lock? Shouldn't all these operations happen on the UI thread?

**Comment by pauldotknopf** [comment]
Looks like Windows tests are segfaulting for some reason.

File: src/native/QmlNet/QmlNet/types/NetReference.cpp (hunk 1 above)
Comment (devmil) [reply src/native/QmlNet/QmlNet/types/NetReference.cpp]: yes, that's right

File: src/net/Qml.Net/Internal/Types/NetReference.cs (hunk 2 above)
Comment (devmil) [reply src/net/Qml.Net/Internal/Types/NetReference.cs]: Just to be sure. An action that results in a NetReference creation might origin in another thread. To be sure not to destroy this counting I introduced the lock

**Comment by devmil** [comment]
Seems this change introduced an unpredictable side effect. Each run some other tests are failing. Sometimes none. Not good...

File: src/net/Qml.Net/Internal/Types/NetReference.cs (hunk 2 above)
Comment (pauldotknopf) [reply src/net/Qml.Net/Internal/Types/NetReference.cs]: Ok.

**Comment by devmil** [comment]
The ObjectTagger unit tests reset the object id database and introduce this side effect.
As we now have the reverse lookup and use this to retrieve the object instances this problem has been made visible

File: src/net/Qml.Net/Internal/Types/NetReference.cs [hunk 3]
```diff
@@ -25,8 +26,11 @@ public object Instance
         {
             get
             {
-                var handle = Interop.NetReference.GetHandle(Handle);
-                return ((GCHandle) handle).Target;
+                if(ObjectIdReferenceTracker.TryGetObjectFor(ObjectId, out var obj))
+                {
+                    return obj;
+                }
+                return null;
```
Comment (pauldotknopf) [inline src/net/Qml.Net/Internal/Types/NetReference.cs]: We should throw an exception detailing a fatal error here. This shouldn't happen, and if it does, many people will die.

We should probably output a message to stdout as well, considering .NET exceptions don't get passed along PInvoke boundaries.

Which brings me to another broader pointer of something we should eventually do (not now). At some point, we should wrap all callbacks with an exception handler, write the exception/stacktrace to stdout, and then throw the error.

This way, we can quickly tell the different between an actual native segfault issue (should be rare), and a .NET error on the reverse PInvoke side of things.

Comment (pauldotknopf) [reply src/net/Qml.Net/Internal/Types/NetReference.cs]: I created #37 for this.

**Comment by pauldotknopf** [comment]
# With GCHandle allocations

``` ini

BenchmarkDotNet=v0.11.0, OS=macOS High Sierra 10.13.6 (17G65) [Darwin 17.7.0]
Intel Core i7-3635QM CPU 2.40GHz (Ivy Bridge), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.301
  [Host] : .NET Core 2.1.1 (CoreCLR 4.6.26606.02, CoreFX 4.6.26606.05), 64bit RyuJIT
  Core   : .NET Core 2.1.1 (CoreCLR 4.6.26606.02, CoreFX 4.6.26606.05), 64bit RyuJIT

Job=Core  Runtime=Core  Force=True  
InvocationCount=1  UnrollFactor=1  

```
| Method |    Mean |    Error |   StdDev |      Gen 0 | Allocated |
|------- |--------:|---------:|---------:|-----------:|----------:|
|    Run | 8.534 s | 0.2023 s | 0.5868 s | 40000.0000 | 112.38 MB |

# Without GCHandle allocations

``` ini

BenchmarkDotNet=v0.11.0, OS=macOS High Sierra 10.13.6 (17G65) [Darwin 17.7.0]
Intel Core i7-3635QM CPU 2.40GHz (Ivy Bridge), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.301
  [Host] : .NET Core 2.1.1 (CoreCLR 4.6.26606.02, CoreFX 4.6.26606.05), 64bit RyuJIT
  Core   : .NET Core 2.1.1 (CoreCLR 4.6.26606.02, CoreFX 4.6.26606.05), 64bit RyuJIT

Job=Core  Runtime=Core  Force=True  
InvocationCount=1  UnrollFactor=1  

```
| Method |    Mean |    Error |   StdDev |      Gen 0 | Allocated |
|------- |--------:|---------:|---------:|-----------:|----------:|
|    Run | 6.973 s | 0.0800 s | 0.0749 s | 34000.0000 | 104.68 MB |

# Comments

The overall speed is 23% faster! I was expected more, but that works! I think even better is the standard deviation. The old way deviated about a .58 seconds from the mean, while the new way only deviated ~.07. To me, that tells me that garbage collection is a lot more efficient and deterministic.

Merging!

**Pull request merged**